src/connection.ts

```diff
@@ -1774,6 +1777,442 @@ class Connection extends EventEmitter {
     }
 
     this.transitionTo(this.STATE.CONNECTING);
+
+    const controller = new AbortController();
+
+    // Abort the connect operation if we reached the connection timeout
+    const connectTimer = setTimeout(() => {
+      controller.abort();
+    }, this.config.options.connectTimeout);
+
+    // Abort the connect operation if the connection is closed
+    const onClose = () => { controller.abort(); };
+    const closeSignal = this.closeController.signal;
+    closeSignal.addEventListener('abort', onClose, { once: true });
+
+    this.establishConnection(controller.signal).finally(() => {
+      clearTimeout(connectTimer);
+      closeSignal.removeEventListener('abort', onClose);
+    }).catch((err) => {
+      if (!closeSignal.aborted && err instanceof Error && err.name === 'AbortError') {
+        const message = `Failed to connect to ${this.config.server}${this.config.options.port ? `:${this.config.options.port}` : `\\${this.config.options.instanceName}`} in ${this.config.options.connectTimeout}ms`;
+        this.debug.log(message);
+
+        err = new ConnectionError(message, 'ETIMEOUT');
+      }
+
+      process.nextTick(() => {
+        this.emit('connect', err);
+        this.transitionTo(this.STATE.FINAL);
+      });
+    });
+  }
+
+  /**
+   * @private
+   */
+  async lookupInstancePort(signal: AbortSignal) {
+    const lookup = new InstanceLookup();
+
+    return (await promisify(lookup.instanceLookup.bind(lookup))({
+      server: this.config.server,
+      instanceName: this.config.options.instanceName!,
+      timeout: this.config.options.connectTimeout,
+      signal: signal
+    }))!;
+  }
+
+  /**
+   * @private
+   */
+  async establishConnection(signal: AbortSignal) {
+    let port;
+    if (this.config.options.port) {
+      port = this.config.options.port;
+    } else {
+      try {
+        port = await this.lookupInstancePort(signal);
+      } catch (err: any) {
+        if (err.name === 'AbortError') {
+          throw err;
+        }
+
+        throw new ConnectionError(err.message, 'EINSTLOOKUP');
+      }
+    }
+
+    let socket;
+    try {
+      socket = await this.connectOnPort(port, this.config.options.multiSubnetFailover, signal);
+    } catch (err: any) {
+      if (err.name === 'AbortError') {
+        throw err;
+      }
+
+      const message = `Failed to connect to ${this.config.server}:${this.config.options.port} - ${err.message}`;
+      this.debug.log(message);
+      throw new ConnectionError(message, 'ESOCKET');
+    }
+
+    socket.setKeepAlive(true, KEEP_ALIVE_INITIAL_DELAY);
+
+    this.messageIo = new MessageIO(socket, this.config.options.packetSize, this.debug);
+    this.messageIo.on('secure', (cleartext) => { this.emit('secure', cleartext); });
+
+    this.socket = socket;
+
+    this.closed = false;
+    this.debug.log('connected to ' + this.config.server + ':' + this.config.options.port);
+
+    this.sendPreLogin();
```

Comment:
Sending of messages currently does not take an `AbortSignal`. It would be important to change this for scenarios where sending data to the server is extremly slow and reaches the connection timeout. Without signal support, the connection would just end up being stuck.

Also, test cases for this would be good.